lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+    | Parraylength _ | Parrayrefu _ | Parraysetu _ | Parrayrefs _ | Parraysets _
+    | Pisint | Pisout
+    | Pignore
+    | Pcompare_ints | Pcompare_floats | Pcompare_bints _
+
+    (* we don't handle array indices as destinations yet *)
+    | (Pmakearray _ | Pduparray _)
+
+    (* we don't handle application primitives as a direct call yet *)
+    | Prevapply | Pdirapply
+
+    (* we don't handle { foo with x = ...; y = recursive-call } *)
+    | Pduprecord _
+
+    | (
+      (* operations returning boxed values could be considered constructions someday *)
+      Pbintofint _ | Pintofbint _
+      | Pcvtbint _
+      | Pnegbint _
+      | Paddbint _ | Psubbint _ | Pmulbint _ | Pdivbint _ | Pmodbint _
+      | Pandbint _ | Porbint _ | Pxorbint _ | Plslbint _ | Plsrbint _ | Pasrbint _
+      | Pbintcomp _
+    )
+    | Pbigarrayref _ | Pbigarrayset _
+    | Pbigarraydim _
+    | Pstring_load_16 _ | Pstring_load_32 _ | Pstring_load_64 _
+    | Pbytes_load_16 _ | Pbytes_load_32 _ | Pbytes_load_64 _
+    | Pbytes_set_16 _ | Pbytes_set_32 _ | Pbytes_set_64 _
+    | Pbigstring_load_16 _ | Pbigstring_load_32 _ | Pbigstring_load_64 _
+    | Pbigstring_set_16 _ | Pbigstring_set_32 _ | Pbigstring_set_64 _ | Pctconst _
+    | Pbswap16
+    | Pbbswap _
+    | Pint_as_pointer
+      ->
+        let primargs = traverse_list ctx primargs in
+        Choice.return (Lprim (prim, primargs, loc))
+
+  and choice_list ctx ~tail terms =
+    Choice.list (List.map (choice ctx ~tail) terms)
+  and choice_pair ctx ~tail (t1, t2) =
+    Choice.pair (choice ctx ~tail t1, choice ctx ~tail t2)
+  and choice_option ctx ~tail t =
+    Choice.option (Option.map (choice ctx ~tail) t)
+
+  in choice ctx t
+
+and traverse ctx = function
+  | Lletrec (bindings, body) ->
+      let ctx, bindings = traverse_letrec ctx bindings in
+      Lletrec (bindings, traverse ctx body)
+  | lam ->
+      shallow_map (traverse ctx) lam
+
+and traverse_letrec ctx bindings =
+  let ctx = List.fold_left declare_binding ctx bindings in
+  let bindings = List.concat_map (traverse_binding ctx) bindings in
+  ctx, bindings
+
+and traverse_binding ctx (var, def) =
+  match find_candidate def with
+  | None -> [(var, traverse ctx def)]
+  | Some lfun ->
+  let special = Ident.Map.find var ctx.specialized in
+  let fun_choice = choice ctx ~tail:true lfun.body in
+  if not fun_choice.Choice.has_tmc_calls then
+    Location.prerr_warning
+      (Debuginfo.Scoped_location.to_location lfun.loc)
+      Warnings.Unused_tmc_attribute;
+  let direct =
+    Lfunction { lfun with body = Choice.direct fun_choice } in
+  let dps =
+    let dst = {
+      var = Ident.create_local "dst";
+      offset = Ident.create_local "offset";
+      loc = lfun.loc;
+    } in
+    let dst_lam = { dst with offset = Lvar dst.offset } in
+    Lambda.duplicate @@ Lfunction { lfun with (* TODO check function_kind *)
+      params = add_dst_params dst lfun.params;
```

Comment:
This might push the number of arguments over the limit for tail calls. I don't know how this should be handled.